pageserver/src/tenant/timeline.rs

```diff
@@ -2709,28 +2713,45 @@ impl Timeline {
                         assert!(!*expect_initdb_optimization, "expected initdb optimization");
                     }
                 }
-                // normal case, write out a L0 delta layer file.
-                let (delta_path, metadata) = self.create_delta_layer(&frozen_layer).await?;
-                HashMap::from([(delta_path, metadata)])
+                // Normal case, write out a L0 delta layer file.
+                // For delta layers, `create_delta_layer` will not modify the layer map.
+                // We will remove frozen layer and add delta layer in one atomic operation later.
+                let layer = self.create_delta_layer(&frozen_layer).await?;
+                (
+                    HashMap::from([(layer.filename(), LayerFileMetadata::new(layer.file_size()))]),
+                    Some(layer),
+                )
             };
 
-        pausable_failpoint!("flush-frozen-before-sync");
-
         // The new on-disk layers are now in the layer map. We can remove the
         // in-memory layer from the map now. The flushed layer is stored in
         // the mapping in `create_delta_layer`.
         {
             let mut guard = self.layers.write().await;
-            let l = guard.layer_map_mut().frozen_layers.pop_front();
 
-            // Only one thread may call this function at a time (for this
-            // timeline). If two threads tried to flush the same frozen
-            // layer to disk at the same time, that would not work.
-            assert!(compare_arced_layers(&l.unwrap(), &frozen_layer));
+            if let Some(ref l) = delta_layer_to_add {
+                // TODO: move access stats, metrics update, etc. into layer manager.
+                l.access_stats().record_residence_event(
+                    &guard,
+                    LayerResidenceStatus::Resident,
+                    LayerResidenceEventReason::LayerCreate,
+                );
 
+                // update metrics
+                let sz = l.file_size();
+                self.metrics.resident_physical_size_gauge.add(sz);
+                self.metrics.num_persistent_files_created.inc_by(1);
+                self.metrics.persistent_bytes_written.inc_by(sz);
+            }
+
+            guard.finish_flush_l0_layer(delta_layer_to_add, &frozen_layer);
             // release lock on 'layers'
         }
 
+        // originally, this pausable failpoint is before the above code block, but we move it here
+        // because it requires the updated data is inside the layer map, after PR#4270 refactor.
+        // probably we can merge these two failpoints.
+        pausable_failpoint!("flush-frozen-before-sync");
```

Comment:
At least for what #4694 fixed (and follow-up in #4725 for the missing comment) I am unsure if the test case anymore makes sense.

The test case was demonstrating that we have this race condition between flushed L0 becoming visible and compaction being able to run *before* we schedule an upload.

If this is now a single lockful operation, perhaps this will now need to go...

However I wonder if scheduling the upload should happen inside this block as well? But alas, there's there are locks.

Reply:
Yeah if the failpoint is moved, the test won't fail any more even if the lock for some reason did not work.

Reply:
I guess the test case you mentioned is `test_compaction_delete_before_upload` in `test_remote_storage.py`? Besides that, there are some other test cases using this fail point, and `test_compaction_delete_before_upload` is using two of them together to wait for some time then exit. Therefore, I renamed these two fail points in https://github.com/neondatabase/neon/pull/4720/commits/fd93a7779ed471a8d3be9112dd853e3803e89396, one for pause, and one for exit.

If we are going to directly remove `test_compaction_delete_before_upload` in this PR, I can make it a single fail point and push another commit.

Reply:
Oh I missed Arpad's comment when I was posting this, then I think it would be good to directly remove this test case.

Reply:
Hmm... But I still think it is possible that the remote client will upload files that are removed by compaction? In this case, we should find another way of reproducing the error instead of removing it, I think... I'll look into that.

Reply:
i.e., the pause fail point should be set in the remote client instead of in compaction, I guess? Also we should force one or two compactions before uploading.

Reply:
If you know of another way to reproduce the issue, it would be really great to have a test for that. In general I'm fine with removal of `test_compaction_delete_before_upload` if the underlying race condition is fixed.

Reply:
I've added the commit https://github.com/neondatabase/neon/pull/4720/commits/5d16be8ebef483080d329709092b11c9b7c1a112 to ensure upload will stuck. This would help reproduce the test case more stably. Let's see if it passes CI.

Reply:
After further investigation the issue can be reproduced without changing the code. The fail point name was wrong in the previous commit and that's why the test does not pass.

This PR make layer map operation atomic, but it didn't make the schedule upload thing in the same critical section as layer map operations. Therefore, the issue still exists.

Anyway, it's now ready for review, and this test case still makes sense. We don't need to remove it.

Reply:
That is to say, moving the fail point still makes the test fail, as long as we add enough sleep between "modify the layer map" and "schedule upload".